### ISSUE TITLE
Docs for Mixed precision training and the use of `LossScaleOptimizer` in custom loops with jax backend

### ISSUE THREAD

**Issue by VachanVY** [keras-team/keras#19244]
Hi,
Should we Scale the loss and gradients like in the [tf.keras docs](https://www.tensorflow.org/guide/mixed_precision#loss_scaling:~:text=%3D%20optimizer.-,get_scaled_loss,-(loss)) for jax backend or the `LossScaleOptimizer` takes care of it in `float16` training?
Tried just wrapping AdamW in it thinking it'll take care of scaling but the training doesn't improve after some steps (but model trained in `float32`)

I checked [Mixed precision policy API](https://keras.io/api/mixed_precision/policy/#dtypepolicy-class), [examples](https://keras.io/examples/), [model.fit to see how it handles it but didn't understand](https://github.com/keras-team/keras/blob/e6e62405fa1b4444102601636d871610d91e5783/keras/backend/tensorflow/trainer.py#L247) and [`LossScaleOptimizer`](https://keras.io/api/optimizers/loss_scale_optimizer/) for mixed precision training in jax but did not find it.
Please include the details in the docs. Thanks

**Comment by VachanVY** [comment]
@SuryanarayanaY Please can you respond??? @sachinprasadhs could you please see this...

**Comment by SuryanarayanaY** [comment]
Hi @VachanVY ,

Are you using custom training loop or simple model.fit. AFAIK without custom training loop model.fit should take care of it. With custom training we need to do some extra steps as mentioned for Tensorflow backend. It seems documentation for this not yet available. I will cross check again and will update.

Thanks!

**Comment by VachanVY** [comment]
I'm using a Fully Custom Training Loop. Checked the Source Code felt like `LossScaleOptimizer` is taking take of scaling and maybe descaling

```python
    def stateless_apply(self, optimizer_variables, grads, trainable_variables):
        if not self.built:
            raise ValueError(
                f"To call `stateless_apply`, {self.__class__.__name__} "
                "must be built (i.e. its variables must have been created). "
                "You can build it via `optimizer.build(trainable_variables)`."
            )
        finite = self.check_finite(grads)
        return ops.cond(
            finite,
            lambda: self._stateless_handle_finite_grads(
                optimizer_variables, grads, trainable_variables
            ),
            lambda: self._stateless_handle_non_finite_grads(
                optimizer_variables, trainable_variables
            ),
        )
```
https://github.com/keras-team/keras/blob/e6e62405fa1b4444102601636d871610d91e5783/keras/optimizers/loss_scale_optimizer.py#L103
```python
    def _stateless_handle_finite_grads(
        self, optimizer_variables, grads, trainable_variables
    ):
        def upscale():
            mapping = list(zip(self.variables, optimizer_variables))
            with backend.StatelessScope(state_mapping=mapping) as scope:
                self.step_counter.assign(0)
                self.dynamic_scale.assign(self.dynamic_scale * 2.0)
            return [scope.get_current_value(v) for v in self._variables]

        def increment():
            mapping = list(zip(self.variables, optimizer_variables))
            with backend.StatelessScope(state_mapping=mapping) as scope:
                self.step_counter.assign_add(1)
            return [scope.get_current_value(v) for v in self._variables]

        mapping = list(zip(self.variables, optimizer_variables))
        with backend.StatelessScope(state_mapping=mapping):
            # Potentially upscale loss and reset counter.
            own_variables = ops.cond(
                ops.equal(self.step_counter, self.dynamic_growth_steps - 1),
                upscale,
                increment,
            )

            # Unscale gradients.
            scale = self.dynamic_scale
            unscaled_grads = [
                g if g is None else ops.divide(g, scale) for g in grads
            ]
            (
                new_trainable_variables,
                new_inner_variables,
            ) = self.inner_optimizer.stateless_apply(
                self.inner_optimizer.variables,
                unscaled_grads,
                trainable_variables,
            )

        new_optimizer_variables = own_variables + new_inner_variables
        return new_trainable_variables, new_optimizer_variables
```

**Comment by VachanVY** [comment]
@SuryanarayanaY Can you please mention atleast briefly about how to use it, it's a bit urgent.... Thanks.

**Comment by VachanVY** [comment]
@SuryanarayanaY It's been a week, could you please reply by telling me how to use it, moreover it's very important for training large models.

**Comment by VachanVY** [comment]
@fchollet Could you please look into this? Thanks.

**Comment by SuryanarayanaY** [comment]
Hi @VachanVY ,

Could you please submit a code snippet on how you have done custom training loop? May be it helps us. Thanks!

**Comment by VachanVY** [comment]
[from my repo](https://github.com/VachanVY/gpt.jax/tree/main)

```python
@partial(jax.jit, static_argnums=-1)
def compute_loss(trainable_variables:list, non_trainable_variables:list, X_batch:Array, y_batch:Array, num_grad_accumalation_steps:int):
    logits, non_trainable_variables = model.stateless_call(
        trainable_variables,  non_trainable_variables,
        X_batch
    )
    loss = loss_fn(y_batch, logits)
    accuracy = get_accuracy(y_batch, logits)
    unscaled_loss = loss/num_grad_accumalation_steps
    scaled_loss = optimizer.scale_loss(unscaled_loss)
    return scaled_loss, (unscaled_loss, accuracy, non_trainable_variables)
grad_fn = jax.value_and_grad(compute_loss, has_aux=True)

@partial(jax.jit, static_argnums=-1)
def mini_step(train_state:Sequence[list], X_batch:Array, y_batch:Array, num_grad_accumalation_steps:int):
    trainable_variables, non_trainable_variables = train_state

    (_, aux), scaled_grad = grad_fn(
        trainable_variables, non_trainable_variables, X_batch, y_batch,
        num_grad_accumalation_steps
    )
    (unscaled_loss, accuracy, non_trainable_variables) = aux
    return scaled_grad, (unscaled_loss, accuracy), (trainable_variables, non_trainable_variables)

@jax.jit
def update_params(grads:list, trainable_variables:list, optimizer_variables:list):
    trainable_variables, optimizer_variables = optimizer.stateless_apply( # also handles scaled grads if LossScalerOPtimizer is used
        optimizer_variables, grads, trainable_variables
    ) # returns updated trainable_variables
    return trainable_variables, optimizer_variables
```

**Comment by dhantule** [comment]
Hi @VachanVY, could you please refer this [similar issue](https://github.com/keras-team/keras/issues/20894).

**Comment by dhantule** [comment]
Hi @VachanVY, thanks for reporting this. 

Please provide standalone code to reproduce this issue. Thanks.

**Comment by VachanVY** [comment]
I was asking for more documentation for it. Not really an issue.
